tests/tests.cpp

```diff
@@ -0,0 +1,398 @@
+
+        // Encrypt a ct with the new key
+        std::vector<uint8_t> new_ct(crypto.AsymEncSize(msg.size()));
+        ASSERT_EQ(0, crypto.AsymEnc(new_key.data(), msg.data(), new_ct.data(),
+                                    msg.size()));
+
+        ASSERT_NE(0, crypto.AsymDec(new_ct.data(), pt.data(), new_ct.size(),
+                                    &_pt_size))
+            << "Ciphertext decrypted successfully with incorrect key";
+
+        // 3) Remove last block of ciphertext
+        std::vector<uint8_t> perturbed_ct(ct);
+        perturbed_ct.resize(perturbed_ct.size() - RSA_MOD_SIZE);
+
+        ASSERT_NE(0, crypto.AsymDec(perturbed_ct.data(), pt.data(),
+                                    perturbed_ct.size(), &_pt_size))
+            << "Truncated ciphertext decrypted successfully";
+
+        // 4) Swap random blocks in the ciphertext
+        for (int i = 0; i < 100; i++) {
+            std::vector<uint8_t> perturbed_ct(ct);
+            swap_blocks(perturbed_ct, RSA_MOD_SIZE);
+            ASSERT_NE(0, crypto.AsymDec(perturbed_ct.data(), pt.data(),
+                                        perturbed_ct.size(), &_pt_size))
+                << "Ciphertext decrypted successfully with swapped blocks";
+        }
+    }
+
+    TEST_F(SoundnessTest, Sig) {
+        // Sign the message
+        std::vector<uint8_t> sig(crypto.AsymSignSize());
+        ASSERT_EQ(0, crypto.Sign(msg.data(), sig.data(), msg.size()));
+
+        // 1) Make random pertubations in the signature
+        for (int i = 0; i < 100; i++) {
+            std::vector<uint8_t> perturbed_sig(sig);
+            perturb(perturbed_sig);
+            ASSERT_NE(0, crypto.Verify(pem_key, msg.data(),
+                                       perturbed_sig.data(), msg.size()))
+                << "Perturbed signature verified successfully";
+        }
+
+        // 2) Make random pertubations in the message
+        for (int i = 0; i < 100; i++) {
+            std::vector<uint8_t> perturbed_msg(msg);
+            perturb(perturbed_msg);
+            ASSERT_NE(0, crypto.Verify(pem_key, perturbed_msg.data(),
+                                       sig.data(), perturbed_msg.size()))
+                << "Perturbed message verified successfully";
+        }
+
+        // 3) Verify with the wrong key
+        // Load public key from file
+        std::vector<uint8_t> new_key;
+        load_file(std::string("../../tests/data/public.pub"), new_key);
+
+        ASSERT_NE(0, crypto.Verify(new_key.data(), msg.data(), sig.data(),
+                                   msg.size()))
+            << "Message verified with the incorrect key";
+    }
+
+    TEST_F(SoundnessTest, Attestation) {
+        // Currently the unittests only support building for the host, so we
+        // can't generate evidence. Instead we modify pre-generated evidence
+
+        // Load nonce, public key, enclave signing key, and evidence from file
```

Comment:
can you add a note here stating that the pre-generated evidence uses this particular nonce, public key, and enclave signing key?

Reply:
Added